include/circt/Dialect/Seq/SeqOps.td

```diff
@@ -220,3 +225,162 @@ def WritePortOp : SeqOp<"write", [
   let results = (outs);
   let hasCustomAssemblyFormat = 1;
 }
+
+//===----------------------------------------------------------------------===//
+// FIRRTL-flavored memory
+//===----------------------------------------------------------------------===//
+
+def FirMemOp : SeqOp<"firmem", [
+  MemoryEffects<[MemAlloc]>,
+  DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]>
+]> {
+  let summary = "A FIRRTL-flavored memory";
+  let description = [{
+    The `seq.firmem` op represents memories lowered from the FIRRTL dialect. It
+    is used to capture some of the peculiarities of what FIRRTL expects from
+    memories, while still representing them at the HW dialect level.
+
+    A `seq.firmem` declares the memory and captures the memory-level parameters
+    such as width and depth or how read/write collisions are resolved. The read,
+    write, and read-write ports are expressed as separate operations that take
+    the declared memory as an operand.
+  }];
+
+  let arguments = (ins
+    I32Attr:$readLatency,
+    I32Attr:$writeLatency,
+    RUWAttr:$ruw,
+    WUWAttr:$wuw,
+    OptionalAttr<StrAttr>:$name,
+    OptionalAttr<SymbolNameAttr>:$inner_sym,
+    OptionalAttr<FirMemInitAttr>:$init,
+    OptionalAttr<StrAttr>:$prefix,
+    OptionalAttr<AnyAttr>:$output_file
+  );
```

Comment:
I think for better PnR implementation experience, should we add optional attribute of hard macro name, and then expose them to repl-seq-mem flow?

Additionally, I think we should add extra IO for MBIST and DFT flow.

Reply:
for simulation, we should add a config to optionally tie down the extra IO since there are two types of simulations for memory:
- the trivial behavior simulation, which the RTL designer only knows the memory size and ports.
- post-RTL simulation should be able to introduce DFT and MBIST in Chisel designs that we should replace these memory with the behavior memory IP from vendor.

Reply:
I could imagine that we add dedicated `seq.mem.mbist_port` and `seq.mem.dft_port` ops on the `seq.mem` op we are trying to distill out of `seq.firmem` later on. But it's not totally clear yo to me how to handle these, since the MBIST/DFT stuff is usually _very_ technology node dependent. Maybe some more generic `seq.mem.opaque_port` with arbitrary operands/results?

I'm starting to think that we should start collecting lots of different memory ops in Seq, covering all the different use cases we have. And then either the commonality becomes very obvious, or the need for an op interface and multiple dedicated ops is the way forward.

Reply:
DFT may be very technology node specific, but can it be treated as an opaque type/bus which is routable, even if it's not decomposable?  Presumably there are generic DFT network building nodes too.

Reply:
Yeah that could work. That way the DFT specifics could be "implementation-defined", or defined by a dedicated DFT pass that runs later. And the memory could just allow this as some opaque port.